Correct input handling for aeif_cond_beta_multisynapse #510 and Update aeif_cond_beta_multisynapse to new aeif numerics approach #514

golosio opened the pull request:
I modified input handling for aeif_cond_beta_multisynapse according to the decision made in the NEST Developer VC 19 Sep 2016, see issue https://github.com/nest/nest-simulator/issues/510

golosio commented:
I also adapted the example in aeif_cond_beta_multisynapse.h and the SLI test.
I still have to do the check that weights are positive and take care of issue https://github.com/nest/nest-simulator/issues/514

golosio commented:
@heplesser about the weight check, should I use a simple assert, as
  assert( e.get_delay() > 0 );
on the same line as the other checks that are done in handle( SpikeEvent& e ), or should I throw an exception with an error message, as
throw BadProperty("Synaptic weights for conductance based models must be positive." );

heplesser commented:
@golosio Raising an exception is certainly more user friendly, so I would suggest that.

golosio commented:
Ok, I should have finished. I made a test using the same parameters of the lsodar test, with one receptor port and synaptic rise time = decay time so that the models behaves as the alpha model, and the output is the same as that of aeif_cond_alpha, see the scripts:
[aeif_ms.py.txt](https://github.com/nest/nest-simulator/files/518172/aeif_ms.py.txt)
[aeif_DT0_ms.py.txt](https://github.com/nest/nest-simulator/files/518173/aeif_DT0_ms.py.txt)
@heplesser and @Silmathoron I wait for your revision.

Silmathoron commented on models/aeif_cond_beta_multisynapse.cpp:
```diff
@@ -144,84 +153,75 @@ aeif_cond_beta_multisynapse::Parameters_::get( DictionaryDatum& d ) const
-        "The neuron has connections, therefore the number of ports cannot be "
-        "reduced." );
+        "The reversal potential, synaptic rise time and synaptic decay time "
+        "arrays must be modified altogether." );
```
I don't think "altogether" is the appropriate term here...
Beyond that, I think what we want to check is only change in length... users should be able to modify only one of the properties as long as they change the values inside and not the number of values, right?

golosio replied:
It's a problem with my English, I will fix the error message.... About the implementation of the constraints, I followed @heplesser suggestions in comment https://github.com/nest/nest-simulator/issues/510#issuecomment-252402610 when he says:
"if a call to SetStatus contains taus_syn, it also must contain E_rev, and they must be of the same length". In case of aeif_cond_alpha_multisynapse there is no taus_syn, but taus_rise and taus_decay, so I guess that SetStatus should either contain none of them or all three...
If I understand well, what you are proposing is that you can also call SetStatus with one or two of them as far as they have the same length that they had previously. For me it would be fine. @heplesser what do you think about this?

golosio replied:
I changed the error message. Please check if it is ok now, otherwise please suggest a suitable one. I modified the behavior for changes in the arrays according to @Silmathoron indication.

heplesser commented on models/aeif_cond_beta_multisynapse.cpp:
```diff
@@ -76,23 +79,28 @@ aeif_cond_beta_multisynapse::Parameters_::Parameters_()
   , b( 80.5 )         // pA
   , V_th( -50.4 )     // mV
-  , I_e( 0.0 )        // pA
+  , num_of_receptors( 1 )
```
I wonder if we should have this as a member. Since it will always be the same as `E_ref.size()`, I think we should should just use
```C++
inline size_t n_receptors() { return E_rev.size(); }
```

heplesser commented on models/aeif_cond_beta_multisynapse.cpp:
```diff
@@ -144,84 +153,75 @@ aeif_cond_beta_multisynapse::Parameters_::get( DictionaryDatum& d ) const
-  if ( updateValue< std::vector< double > >( d, names::taus_decay, tau_tmp ) )
-  {
-    if ( tau_tmp.size() < taus_decay.size() && has_connections_ == true )
+  std::vector< double > Erev_tmp, taur_tmp, taud_tmp;
```
You should avoid defining multiple variables in one statement.

You also don't need temporary variables here, since `set()` is called on a temporary copy of `P_` that will only be copied back to the real `P_` if no exception is raised. This may also simplify some of the error checking below.

golosio replied:
removed.

heplesser commented on models/aeif_cond_beta_multisynapse.cpp:
```diff
@@ -144,84 +153,75 @@ aeif_cond_beta_multisynapse::Parameters_::get( DictionaryDatum& d ) const
+    taus_rise = taur_tmp;
+    taus_decay = taud_tmp;
+    num_of_receptors = taur_tmp.size();
+    receptor_types.resize( num_of_receptors );
```
See above, I think we can drop `receptor_types`.

golosio replied:
done

heplesser commented on models/aeif_cond_beta_multisynapse.cpp:
```diff
@@ -656,44 +686,37 @@ aeif_cond_beta_multisynapse_dynamics( double,
   const double& w = y[ S::W ];
 
   double I_syn = 0.0;
 
```
I would remove this empty line to tie the initialization more directly to the loop.

golosio replied:
done

heplesser commented on models/aeif_cond_beta_multisynapse.cpp:
```diff
@@ -656,44 +686,37 @@ aeif_cond_beta_multisynapse_dynamics( double,
-    {
-      I_syn += g * ( V - node.P_.E_in ); // g<0, V>E_in => I_syn decreases
-    }
+    size_t j = i * S::NUMBER_OF_STATES_ELEMENTS_PER_RECEPTOR;
```
Add `const`.

golosio replied:
done

heplesser commented on models/aeif_cond_beta_multisynapse.cpp:
```diff
@@ -656,44 +686,37 @@ aeif_cond_beta_multisynapse_dynamics( double,
-      I_syn += g * ( V - node.P_.E_in ); // g<0, V>E_in => I_syn decreases
-    }
+    size_t j = i * S::NUMBER_OF_STATES_ELEMENTS_PER_RECEPTOR;
+    // I_syn = I_syn + g*(E_rev - V)
```
Maybe move this comment above the `I_syn = 0` initialization and make the summation explicit: `I_syn = - sum_k g_k (V - E_rev_k)`.

heplesser commented on models/aeif_cond_beta_multisynapse.cpp:
```diff
@@ -656,44 +686,37 @@ aeif_cond_beta_multisynapse_dynamics( double,
-  for ( size_t i = 0; i < node.P_.num_of_receptors_; ++i )
+  for ( size_t i = 0; i < node.P_.num_of_receptors; ++i )
   {
     size_t j = i * S::NUMBER_OF_STATES_ELEMENTS_PER_RECEPTOR;
```
`const`

golosio replied:
done

heplesser commented on models/aeif_cond_beta_multisynapse.h:
```diff
@@ -74,16 +74,18 @@
  neuron = nest.Create('aeif_cond_beta_multisynapse')
  nest.SetStatus(neuron, {"V_peak": 0.0, "a": 4.0, "b":80.5})
- nest.SetStatus(neuron, {'taus_decay':[50.0,20.0,20.0,20.0],
+ nest.SetStatus(neuron, {'num_of_receptors':4,
```
`num_of_receptors` should never be set explicitly.

golosio replied:
removed

heplesser commented on models/aeif_cond_beta_multisynapse.h:
```diff
@@ -74,16 +74,18 @@
+ w=[1.0, 1.0, 1.0, 1.0]
  for syn in range(4):
      nest.Connect(spike, neuron, syn_spec={'model': 'static_synapse',
                                            'weight': w[syn],
```
Don't you need to specify `receptor_type` here?

golosio replied:
You mean that the order should be important? I do not know, I did not see any difference in the simulations if I put 'receptor_type' at the end of the syn_spec list, in any case now I've put it in the second position after 'model'.

heplesser commented on testsuite/unittests/test_aeif_cond_beta_multisynapse.sli:
```diff
@@ -75,12 +75,14 @@ if
-/V_peak			0. def
-/a 				4. def
-/b 				80.5 def
+/V_peak		0. def
```
Fix whitespace here and in the following lines.

golosio replied:
done

heplesser commented on testsuite/unittests/test_aeif_cond_beta_multisynapse.sli:
```diff
@@ -75,12 +75,14 @@ if
+/V_peak		0. def
+/a 		4. def
+/b 		80.5 def
+/num_of_receptors       3 def
```
Should not be set explicitly.

golosio replied:
removed

heplesser commented on testsuite/unittests/test_aeif_cond_beta_multisynapse.sli:
```diff
@@ -75,12 +75,14 @@ if
 /taus_decay		[400.0 200.0 200.0] def % synaptic decay times  
 /taus_rise		[200.0 100.0 100.0] def % synaptic rise times  
-/weights		[0.1 0.1 -0.1] def % synaptic weights
+/weights		[0.1 0.1 0.1] def % synaptic weights
```
Wouldn't it be better to make all reversal potentials, time constants and weights different from each other to make sure nothing gets mixed up in the code?

golosio replied:
ok I used values different from each other.

heplesser commented on testsuite/unittests/test_aeif_cond_beta_multisynapse.sli:
```diff
@@ -75,12 +75,14 @@ if
 /taus_rise		[200.0 100.0 100.0] def % synaptic rise times  
-/weights		[0.1 0.1 -0.1] def % synaptic weights
+/weights		[0.1 0.1 0.1] def % synaptic weights
 /delays     	[1. 1000. 1500.] def % ms - synaptic delays
```
Do you need such long delays? Shorter simulation times would make the test run faster.

golosio replied:
ok I increased g_L so the approximation used for the analytical calculation is also valid for smaller synaptic times, and I could also decrease the tolerance

golosio commented on models/aeif_cond_beta_multisynapse.cpp:
```diff
@@ -20,6 +20,8 @@
  *
  */
 
+#include <stdio.h>
```
Ops... sorry I use it for myself for debugging and I forgot it. Removed.

golosio commented on models/aeif_cond_beta_multisynapse.cpp:
```diff
@@ -41,6 +43,7 @@
 #include "doubledatum.h"
 #include "integerdatum.h"
 
+using namespace std;
```
removed

golosio commented on models/aeif_cond_beta_multisynapse.cpp:
```diff
@@ -76,23 +79,28 @@ aeif_cond_beta_multisynapse::Parameters_::Parameters_()
 {
+  static const double E_rev_def = 0.0;       // E_rev default value
+  static const double taus_rise_def = 2.0;   // taus_rise default value
+  static const double taus_decay_def = 20.0; // taus_decay default value
```
good, now the code is much more compact, there is only the initializer list and the body is empty. For now I'm still using the default with one receptor port, but it will be easy to remove it.

golosio commented on models/aeif_cond_beta_multisynapse.cpp:
```diff
@@ -76,23 +79,28 @@ aeif_cond_beta_multisynapse::Parameters_::Parameters_()
+  receptor_types.clear();
+  E_rev.clear();
   taus_rise.clear();
   taus_decay.clear();
```
removed.

golosio commented on models/aeif_cond_beta_multisynapse.cpp:
```diff
@@ -76,23 +79,28 @@ aeif_cond_beta_multisynapse::Parameters_::Parameters_()
   taus_decay.clear();
-  taus_decay.push_back( 20.0 );
-  taus_rise.push_back( 2.0 );
+  receptor_types.push_back( 1 );
```
Right, it was not used anywhere, I removed it.

golosio commented:
@heplesser @Silmathoron ok I made all the changes that you requested.

heplesser commented on models/aeif_cond_beta_multisynapse.cpp:
```diff
@@ -457,15 +439,9 @@ aeif_cond_beta_multisynapse::calibrate()
-  }
-
-  V_.g0_.resize( P_.num_of_receptors );
+  V_.g0_.resize( P_.n_receptors() );
```
I think it would be good for readability to point out that `g0_` will be initialized in the loop below.

heplesser commented on models/aeif_cond_beta_multisynapse.cpp:
```diff
@@ -501,10 +477,9 @@ aeif_cond_beta_multisynapse::calibrate()
   S_.y_.resize( State_::NUMBER_OF_FIXED_STATES_ELEMENTS
-    + ( State_::NUMBER_OF_STATES_ELEMENTS_PER_RECEPTOR
-                  * P_.num_of_receptors ) );
+    + ( State_::NUMBER_OF_STATES_ELEMENTS_PER_RECEPTOR * P_.n_receptors() ) );
```
I think it would be good to explicitly give the initial value `0` as second argument to `resize()`.

heplesser commented on models/aeif_cond_beta_multisynapse.cpp:
```diff
@@ -168,60 +156,54 @@ aeif_cond_beta_multisynapse::Parameters_::set( const DictionaryDatum& d )
   updateValue< double >( d, names::g_L, g_L );
 
-  std::vector< double > Erev_tmp, taur_tmp, taud_tmp;
+  size_t old_n_receptors = n_receptors();
```
Should be const.

golosio replied:
done

heplesser commented on models/aeif_cond_beta_multisynapse.cpp:
```diff
@@ -168,60 +156,54 @@ aeif_cond_beta_multisynapse::Parameters_::set( const DictionaryDatum& d )
   bool taud_flag =
-    updateValue< std::vector< double > >( d, names::taus_decay, taud_tmp );
+    updateValue< std::vector< double > >( d, names::taus_decay, taus_decay );
   if ( Erev_flag || taur_flag || taud_flag )
```
I think it should be possible to simplify this some more:
```C++
// n_receptors() == E_rev.size()
if ( n_receptors() != taus_rise.size() || n_receptors() != taus_decay.size() )
{
  throw BadProperty("E_rev, taus_raus and taus_decays must have the same size.");
}
```
Now we know all is consistent I would then check `n_receptors() != old_n_receptors && has_connections` and `n_receptors > 0`. Then the check that the taus > 0 and tau_rise < tau_decay. That should maybe be protected by `if ( taur_flag || taud_flag )` since we have loops here and would want to avoid running through them if the taus have not been changed.

golosio replied:
I agree that the code can be made more compact if we simply want to check consistency. However my idea is that we should not only check for consistency, but also give the user a feedback that suggests him how he should modify his command. For instance, if the current number of receptor ports is 4 (or if they have not yet been initialized so the number is 0 or 1) and the user types a command as:
nest.SetStatus(neuron, {'taus_decay':[50.0,20.0,20.0]})
I think a message as:
"If the number of receptor ports is changed, all three arrays E_rev, taus_rise and taus_decay must be provided."
is more helpful to the user than:
"E_rev, taus_rise and taus_decay must have the same size."
because with the latter message it may not be obvious to all users that the problem can be solved by providing all three arrays as arguments.

heplesser commented on models/aeif_cond_beta_multisynapse.cpp:
```diff
@@ -168,60 +156,54 @@ aeif_cond_beta_multisynapse::Parameters_::set( const DictionaryDatum& d )
       throw BadProperty( "The neuron must have at least one port." );
     }
-    if ( taur_tmp.size() < num_of_receptors && has_connections_ == true )
+    if ( taus_rise.size() < old_n_receptors && has_connections_ == true )
```
Avoid comparisons with `true` and `false`.

golosio commented:
Ok I'll implement the changes that you suggest. Just one note and a question. When you say

>> It is a bit unusual that State_::set() allows setting of conductances and their derivatives. We don't support that for other models and I am a bit sceptical about adding it here, especially the derivative.

Actually this is  how it is done in aeif_cond_alpha_multisynapse, which was the base for this model; I do not know about other models. Where do you suggest to move this part of the code?

Silmathoron commented:
@golosio I think what @heplesser means is that maybe the user should not be able to modify these variables.
I must say that I do not really agree on that, but I guess we should discuss it on Monday (actually I'm waiting for the VC before I start my review, which is why I haven't been very active lately; I hope all these issues will be clearer afterwards).
In the meantime, I guess you should leave that as it is until then ;)

golosio commented on models/aeif_cond_beta_multisynapse.h:
```diff
@@ -202,16 +201,18 @@ class aeif_cond_beta_multisynapse : public Archiving_Node
 
     void get( DictionaryDatum& ) const; //!< Store current values in dictionary
     void set( const DictionaryDatum& ); //!< Set values from dictionary
+    inline size_t
```
done

golosio commented:
@heplesser I made the changes that you suggested, except for the two for which I wait for your feedback. I also wait for @Silmathoron review after the VC.

Silmathoron commented:
Ok, the VC concluded that default behaviour should be that of the single synapse model, so you should implement a default with 1 synapse and the same parameters as the aeif_cond_alpha, except eventually for the synaptic time constants... I'll let you judge on that unless @heplesser has specific suggestions.
As for the ``dynamics`` functions, the VC agreed on removing the ``_DT0`` so you can update the function according to your suggestion and @heplesser['s comment](https://github.com/nest/nest-simulator/issues/514#issuecomment-253451402).
Eventually, I think the people in the VC mostly thought giving access to all state variables in set was rather uninteresting but no one seemed stronlgy opposed to it either... so my guess is you can do as you feel like, but maybe @heplesser would like to comment on that...

Silmathoron commented:
Ok, there are a few problems that were missed previously on lines [303](https://github.com/nest/nest-simulator/pull/516/files#diff-230597c8da565f8ce73398a1bd42067fR303) and [310](https://github.com/nest/nest-simulator/pull/516/files#diff-230597c8da565f8ce73398a1bd42067fR310):
unless I'm mistaken, ``dg`` is the derivative, so error message on L303 should be ``"Conductances and derivatives arrays should have the same size"``, and on L310, ``dg`` should be allowed to be negative.

Silmathoron commented on models/aeif_cond_beta_multisynapse.cpp:
```diff
@@ -425,6 +422,15 @@ aeif_cond_beta_multisynapse::init_buffers_()
+  // set the right threshold depending on Delta_T
+  if ( P_.Delta_T > 0. )
+  {
+    B_.V_peak = P_.V_peak_;
```
I'm almost sure this is supposed to be in calibrate since we want this check to be done every time ``Simulate`` is called, so ``V_peak`` should be in ``V_``; @heplesser, could you confirm?

golosio replied:
done

Silmathoron commented on models/aeif_cond_beta_multisynapse.cpp:
```diff
@@ -656,46 +662,30 @@ aeif_cond_beta_multisynapse_dynamics( double,
 
   // shorthand for state variables
-  const double& V = y[ S::V_M ];
+  const double& V = std::min( y[ S::V_M ], node.P_.V_peak_ ); // bound V
```
Maybe we should add a comment to assure people that we indeed want to use ``P_.V_peak_`` and not ``V_.V_peak`` here...

golosio replied:
done. However when I will commit please check if this is sufficient or please suggest a more suitable comment.

Silmathoron commented on models/aeif_cond_beta_multisynapse.cpp:
```diff
@@ -563,7 +563,7 @@ aeif_cond_beta_multisynapse::update( Time const& origin,
         S_.y_[ State_::V_M ] = P_.V_reset_;
       }
-      else if ( S_.y_[ State_::V_M ] >= P_.V_peak_ )
+      else if ( S_.y_[ State_::V_M ] >= B_.V_peak )
```
Switch ``V_peak`` to ``V_`` (see above)

golosio replied:
done

Silmathoron commented on models/aeif_cond_beta_multisynapse.cpp:
```diff
@@ -656,46 +662,30 @@ aeif_cond_beta_multisynapse_dynamics( double,
-  // If the argument is too large, we clip it.
-  const double I_spike =
-    node.P_.Delta_T * std::exp( std::min( exp_arg, MAX_EXP_ARG ) );
+  const double I_spike = node.P_.Delta_T == 0. ? 0 : node.P_.Delta_T
```
To be consistent (and homogeneous), change to ``node.P_.Delta_T * node.P_.g_L * std::exp( ( V - node.P_.V_th ) / node.P_.Delta_T )``

heplesser replied:
I think for clarity, it would be good to place everything after the `:` in parentheses, even if not strictly necessary.

Silmathoron commented on models/aeif_cond_beta_multisynapse.cpp:
```diff
@@ -656,46 +662,30 @@ aeif_cond_beta_multisynapse_dynamics( double,
 
   // dv/dt
-  f[ S::V_M ] = ( -node.P_.g_L * ( ( V - node.P_.E_L ) - I_spike ) + I_syn - w
+  f[ S::V_M ] = ( -node.P_.g_L * ( V - node.P_.E_L - I_spike ) + I_syn - w
```
Switch to ``( -node.P_.g_L * ( V - node.P_.E_L ) + I_spike + I_syn + ...``

golosio replied:
done

Silmathoron commented on models/aeif_cond_beta_multisynapse.h:
```diff
@@ -294,6 +298,7 @@ class aeif_cond_beta_multisynapse : public Archiving_Node
      * the first simulation, but not modified before later Simulate calls.
      */
     double I_stim_;
+    double V_peak;
```
Should be in ``V_`` because it can be modified before ``Simulate``

golosio replied:
done

golosio commented:
Thank you @Silmathoron , I implemented the default with one synapse and the same parameters as the aeif_cond_alpha, except for the synaptic rise time and decay time. I updated the dynamic function according to @heplesser suggestion. 
About State_::set() allowing to set conductances and their derivatives, I think @heplesser already said his opinion, so I would remove this possibility, unless you think there is really a reason to allow it

Silmathoron commented:
I would rather say that I don't really see a reason to remove it if you implemented it ^^
Otherwise I don't really care since I don't have a use case for the moment... it's up to you, as far as I am concerned.

golosio commented:
@Silmathoron 
ok since @heplesser was skeptic about it I will remove it.

>> unless I'm mistaken, dg is the derivative, so error message on L303 should be "Conductances and derivatives arrays should have the same size", and on L310, dg should be allowed to be negative.

As I will remove this part of the code, those changes are not necessary any more. However, note that what is called dg in the code IS NOT the derivative of g. If you look for instance Roth and van Rossum, "Modeling synapses"
http://homepages.inf.ed.ac.uk/mvanross/reprints/roth_mvr_chap.pdf
equations 6.7 - 6.9 and compare them with the equations in aeif_cond_beta_multisynapse_dynamic you can observe that what is called dg (or y[ S::DG + j ] ) in the code is actually the function that is called h in those equations, and from equation 6.8 you can see that the derivative dg/dt is not h, because there is another term. Sorry, but the code of aeif_cond_beta_multisynapse was originally based on aeif_cond_alpha_multisynapse, and I did not modify the name of the variable dg although I thought it was somehow inappropriate.

Silmathoron commented:
Ok, I did not see that ``S::DG`` and ``dg`` where different; this is indeed not optimal and speaks in favour of removing it.
By the way, I just noticed that the docstring of the model in ``aeif_cond_beta_multisynapse.h`` does not include a description of the parameters, so could you add one? (see example in [``aeif_cond_alpha``](https://github.com/golosio/nest-simulator/blob/05eef9ae2f5214a7ea3dda967578aba7c03c0e67/models/aeif_cond_alpha.h#L65)

golosio commented:
>> the docstring of the model in aeif_cond_beta_multisynapse.h does not include a description of the parameters, so could you add one? (see example in aeif_cond_alpha

sure, thank you for indicating the example!

golosio commented on models/aeif_cond_beta_multisynapse.cpp:
```diff
@@ -144,84 +141,69 @@ aeif_cond_beta_multisynapse::Parameters_::get( DictionaryDatum& d ) const
-      }
+      throw BadProperty(
+        "The reversal potential, synaptic rise time and synaptic decay time "
+        "arrays must have the same size." );
```
About this point, please have a look at my answer in
https://github.com/nest/nest-simulator/pull/516#discussion_r83532267
I agree that if we simply want to check  that the changes are consistent, the code can be made more compact. However I think that we should also think about what error messages would be more helpful for the user.
I think we should distinguish those two erroneous situations:
1) The user types a command that provides only one or two arrays, and they have a different size from the old one. Then the error message should be something like:
"If the number of receptor ports is changed, all three arrays  E_rev, taus_rise and taus_decay must be provided."
2) The user types a command that provides three arrays, but they have inconsistent sizes
Then the error message should be something like:
"The reversal potential, synaptic rise time and synaptic decay time arrays must have the same size."
If we first agree on which error messages should be printed in different situations, then we can discuss on what is the best way to implement it in the code.

Silmathoron replied:
My bad, I remember reading about something like what you said. Ok, I understand your point.

Silmathoron commented on models/aeif_cond_beta_multisynapse.cpp:
```diff
@@ -233,14 +215,29 @@ aeif_cond_beta_multisynapse::Parameters_::set( const DictionaryDatum& d )
   updateValue< double >( d, names::gsl_error_tol, gsl_error_tol );
 
-  if ( V_peak_ <= V_th )
+  if ( V_reset_ >= V_peak_ )
```
Ok, then this is probably my fault ^^"
No, just change it here, I'll correct the ``aeif_cond_alpha`` it in the PR suppressing the ``dynamics_DT0``

golosio commented:
@Silmathoron Thank you for your review. I think I'm done in making the changes that you requested.

Silmathoron commented on models/aeif_cond_beta_multisynapse.h:
```diff
@@ -64,6 +64,48 @@
  During connection, the ports are selected with the property "receptor_type".
+The membrane potential is given by the following differential equation:
+C dV/dt= -g_L(V-E_L)+g_L*Delta_T*exp((V-V_T)/Delta_T)-g_e(t)(V-E_e)
+                                                     -g_i(t)(V-E_i)-w +I_e
```
Ok, sorry, one last comment: since it's multisynapse, here, you should probably detail:

    C dV/dt= -g_L(V-E_L) + g_L*Delta_T*exp((V-V_T)/Delta_T) + I_syn_tot(V, t) - w + I_e
    
    where:
    I_syn_tot(V,t) = \sum_i g_i(t) (V - E_{rev,i})

    and the synapse i is excitatory or inhibitory depending on the value of E_{rev,i}.

golosio commented:
@Silmathoron done!

heplesser commented on nestkernel/nest_names.h:
```diff
@@ -407,6 +407,8 @@ extern const Name taus_eta;   //!< Specific to population point process model
 extern const Name taus_syn;   //!< Synapse time constants (array)
 extern const Name taus_rise;  //!< Synapse rise constants (array)
 extern const Name taus_decay; //!< Synapse decay constants (array)
+extern const Name E_rev;      //!< Reversal potential (array)
```
Entries in `nest_names.[h,cpp]` should be in strictly alphabetical order.

heplesser commented on nestkernel/nest_names.h:
```diff
@@ -407,6 +407,8 @@ extern const Name taus_eta;   //!< Specific to population point process model
 extern const Name taus_rise;  //!< Synapse rise constants (array)
 extern const Name taus_decay; //!< Synapse decay constants (array)
+extern const Name E_rev;      //!< Reversal potential (array)
+extern const Name num_of_receptors; //!< number of receptor ports
```
In other names, we have used `n_`, e.g., `n_events`.  I suggest `n_receptors` here for consistency.

heplesser commented on models/aeif_cond_beta_multisynapse.cpp:
```diff
@@ -473,14 +470,24 @@ aeif_cond_beta_multisynapse::calibrate()
   S_.y_.resize( State_::NUMBER_OF_FIXED_STATES_ELEMENTS
-    + ( State_::NUMBER_OF_STATES_ELEMENTS_PER_RECEPTOR
-                  * P_.num_of_receptors_ ) );
+    + ( State_::NUMBER_OF_STATES_ELEMENTS_PER_RECEPTOR * P_.n_receptors() ) );
```
I'd suggest explicitly initializing with `0.0` here, as second argument.

Silmathoron commented:
By the way, @golosio, I think you also forgot to readd the ``V_peak_ > V_th`` check in your branch... I was looking for it in your code to add it again in the psc models x)

golosio commented:
@Silmathoron maybe you are looking to an updated version of the file... it should be in line 218.

Silmathoron commented:
My bad, you're probably right, must have been an outdated file.
All good!

heplesser commented:
@golosio As in #513 , I would suggest we allow V_th == V_peak and rephrase the error message.

golosio commented:
@heplesser done!

heplesser commented on models/aeif_cond_beta_multisynapse.h:
```diff
@@ -51,20 +51,66 @@
+ rise time and decay time constants. Synaptic conductance is modeled by a
  beta function, as described by A. Roth and M.C.W. van Rossum
  in Computational Modeling Methods for Neuroscientists, MIT Press 2013,
  Chapter 6
```
Fullstop missing at end of sentence.

heplesser commented on models/aeif_cond_beta_multisynapse.h:
```diff
@@ -51,20 +51,66 @@
- are then automatically assigned and there range is from 1 to n.  (n
- being the index of the last element of the tau_ex and tau_in
- arrays).
+ The number of receptor ports is supplied by the function "n_receptors".
```
I would drop this sentence, and `n_receptors` is not a function. I think the meaning of n_receptors becomes clear from the last sentence in this paragraph.

golosio replied:
done.

heplesser commented on models/aeif_cond_beta_multisynapse.h:
```diff
@@ -51,20 +51,66 @@
+
+ and the synapse i is excitatory or inhibitory depending on the value of
+ E_{rev,i}.
+
```
The differential equation for w fell out in one of your previous commits. I think it should be here. Would you add it again?

golosio replied:
done.

heplesser commented on models/aeif_cond_beta_multisynapse.h:
```diff
@@ -51,20 +51,66 @@
+
  Examples:
  % PyNEST example, of how to assign synaptic rise time and decay time
  % to a receptor type.
```
I think the comment here is not needed. If you want to keep it, mark the comment with Python's `#` instead of `%`, which introduces comments in SLI.

golosio replied:
removed.

golosio commented:
@heplesser ok I made the changes that you requested!

heplesser commented:
@golosio Thank you for your efforts! Together with @Silmathoron's "all good" above, I will merge.

golosio commented:
Thank you @heplesser and @Silmathoron !
